### PR TITLE
Add metadynamics

### PR DESCRIPTION
**Pull request by vitartas**
Integrated PLUMED to mlp-train via ASE calculator. Still need to include Metadynamics methods for data analysis.

### CODE REVIEW

File: mlptrain/log.py [hunk 1]
```diff
@@ -10,9 +10,6 @@
 # Try and use colourful logs
 try:
     import coloredlogs
-    FORMAT = '%(asctime)s %(hostname)s %(name)s[%(processName)s] ' \
```
Comment (tristan-j-wood) [inline mlptrain/log.py:13]: What was the reason for removing this again?

File: mlptrain/sampling/md.py [hunk 2]
```diff
@@ -24,7 +26,8 @@ def run_mlp_md(configuration: 'mlptrain.Configuration',
                init_temp:     Optional[float] = None,
                fbond_energy:  Optional[dict] = None,
                bbond_energy:  Optional[dict] = None,
-               bias:          Optional['mlptrain.bias.Bias'] = None,
+               bias:          Optional[Union['mlptrain.Bias',
+                                             'mlptrain.PlumedBias']] = None,
                method:        Optional[str] = None,
```
Comment (tristan-j-wood) [inline mlptrain/sampling/md.py:31]: I would only allow valid options. At the moment people can type anything in and it won't immediately break if it isn't implemented, which might lead to confusion over what is breaking. Also for strings, lower() so people can type in any case

File: mlptrain/sampling/md.py [hunk 3]
```diff
@@ -99,17 +105,39 @@ def append_energy(_atoms=ase_atoms):
...
+        dyn = VelocityVerlet(ase_atoms, dt_ase)
 
     dyn.attach(append_energy, interval=interval)
     dyn.attach(traj.write, interval=interval)
 
     logger.info(f'Running {n_steps:.0f} steps with a timestep of {dt} fs')
-    dyn.run(steps=n_steps)
+
+    if isinstance(bias, PlumedBias):
+        from ase.calculators.plumed import Plumed
```
Comment (tristan-j-wood) [inline mlptrain/sampling/md.py:120]: Can add some logs saying it is running plumed

File: mlptrain/sampling/md.py [hunk 4]
```diff
@@ -99,17 +105,39 @@ def append_energy(_atoms=ase_atoms):
...
+                             timestep=dt_ase,
+                             atoms=ase_atoms,
+                             kT=temp * ase_units.kB)
+        ase_atoms.calc = plumed_calc
+
+        dyn.run(steps=n_steps)
+        plumed_calc.plumed.finalize()
+
+    else:
+        ase_atoms.calc = mlp.ase_calculator
```
Comment (tristan-j-wood) [inline mlptrain/sampling/md.py:137]: Logs saying it is running ase

File: mlptrain/sampling/md.py [hunk 5]
```diff
@@ -227,3 +255,43 @@ def _n_simulation_steps(dt: float,
...
+        metad_setup = ['METAD '
+                       f'ARG={bias.cv_sequence} '
+                       f'PACE={bias.pace} '
+                       f'HEIGHT={bias.height} '
+                       f'SIGMA={bias.width} '
+                       f'BIASFACTOR={bias.biasfactor} '
+                       f'FILE=HILLS_pid{os.getpid()}.dat']
+        setup.extend(metad_setup)
+
+    # Print
```
Comment (tristan-j-wood) [inline mlptrain/sampling/md.py:289]: Slightly more informative comment might be useful

File: mlptrain/sampling/metadynamics.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+
+        start_metadynamics = time.perf_counter()
+
+        self.temp = temp
+        self.bias.set_metad_params(width=width,
+                                   pace=pace,
+                                   height=height,
+                                   biasfactor=biasfactor)
+
+        metad_processes, metad_trajs = ([] for _ in range(2))
```
Comment (tristan-j-wood) [inline mlptrain/sampling/metadynamics.py:118]: I generally don't use this for small numbers of lists. It's longer and more difficult to read than simply: = [], []

File: mlptrain/sampling/metadynamics.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+                                                       dt,
+                                                       self.bias),
+                                                 kwds=kwargs)
+                metad_processes.append(metad_process)
+
+            for metad_process in metad_processes:
+                metad_trajs.append(metad_process.get())
+
+        if save_sep:
+            os.mkdir('metad_trajectories')
```
Comment (tristan-j-wood) [inline mlptrain/sampling/metadynamics.py:147]: What happens if the directory exists?

File: mlptrain/sampling/plumed.py [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+class PlumedCV:
+    """Parent class containing methods for initialising PLUMED collective
+    variables"""
+
+    def __init__(self,
+                 name: str = None,
+                 atom_groups: Sequence = None,
+                 file_name: str = None):
+        """
+        This class contains methods to initialise PLUMED collective variables
```
Comment (tristan-j-wood) [inline mlptrain/sampling/plumed.py:85]: If it shouldn't be called directly, would it be better to make it a private class?

File: mlptrain/sampling/plumed.py [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+
+        _names = [line.split(':')[0] for line in self.setup]
+        _actions = [line.split(' ')[1] for line in self.setup]
+        _names_and_actions = list(zip(_names, _actions))
+
+        self.name = _names_and_actions.pop()[0]
+
+        for name, action in _names_and_actions:
+            if action != 'GROUP':
+                self.dof_names.append(f'{self.name}_{name}')
```
Comment (tristan-j-wood) [inline mlptrain/sampling/plumed.py:146]: Need to return None

**Review by tristan-j-wood — COMMENTED** [review]
Very nice code

File: mlptrain/log.py (hunk 1 above)
Comment (vitartas) [reply mlptrain/log.py:13]: I've decided to return to using PIDs in log as I've used PIDs in naming PLUMED output files when multiple metadynamics runs execute in parallel.

**Comment by vitartas** [comment]
Adding new commits:
1. Attempted to fix the code based on your comments
2. Contracted the method variable in run_mlp_md() to mlptrain.Bias and mlptrain.PlumedBias as the variable is not supposed to be changed by the user
3. Added a feature to fully initialise mlptrain.PlumedBias using a full PLUMED input file (which allows to execute run_mlp_md with an arbitrary PLUMED input)
4. Added an option to specify the component of the collective variable to use as the actual CV when initialised using mlptrain.PlumedCustomCV

**Review by tristan-j-wood — COMMENTED** [review]
Few cosmetic changes

**Comment by vitartas** [comment]
Adding new commits:
1. Refactored run_mlp_md()
2. Added a restart option for run_mlp_md() and run_metadynamics()
3. Added methods estimate_width() and try_multiple_biasfactors() which are helpful for choosing metadynamics parameters
4. Added plot_fes_convergence() which is useful for checking whether the FES has converged
5. Added compute_fes() and plot_fes()

File: mlptrain/sampling/md.py [hunk 10]
```diff
@@ -1,42 +1,48 @@
...
+               temp:               float,
+               dt:                 float,
+               interval:           int,
+               init_temp:          Optional[float] = None,
+               fbond_energy:       Optional[dict] = None,
+               bbond_energy:       Optional[dict] = None,
+               bias:               Optional = None,
+               restart_files:      Optional[List[str]] = None,
+               copied_substrings:  Sequence[str] = ('.xml', '.json', '.pth'),
+               kept_substrings:    Sequence[str] = None,
```
Comment (tristan-j-wood) [inline mlptrain/sampling/md.py:31]: Add arg for plumed print setup with docstring = False

File: mlptrain/sampling/metadynamics.py [hunk 11]
```diff
@@ -341,41 +325,64 @@ def run_metadynamics(self,
             for metad_process in metad_processes:
                 metad_trajs.append(metad_process.get())
 
-        if save_sep:
-            for idx, metad_traj in enumerate(metad_trajs):
-                metad_traj.save(filename=f'metad_{idx+1}.xyz')
+        self._move_and_save_files(metad_trajs, save_sep, all_to_xyz, restart)
```
Comment (tristan-j-wood) [inline mlptrain/sampling/metadynamics.py:330]: Could have a comment explaining where files are being moved

File: mlptrain/sampling/plumed.py [hunk 12]
```diff
@@ -434,12 +437,13 @@ def __init__(self,
...
 
 
-def plot_cv(filename:    str,
-            time_units:  str = 'ps',
-            cv_units:    Optional[str] = None,
-            cv_limits:   Optional[Sequence[float]] = None,
-            label:       Optional[str] = None
-            ) -> None:
+def plot_cv_versus_time(filename:    str,
+                        style:       str = 'scatter',
```
Comment (tristan-j-wood) [inline mlptrain/sampling/plumed.py:504]: Default could/_should_ be plot

File: mlptrain/sampling/plumed.py [hunk 13]
```diff
@@ -215,6 +216,68 @@ def dof_sequence(self) -> str:
...
+                          ) -> None:
+        """
+        Attaches lower wall bias to the collective variable.
+
+        -----------------------------------------------------------------------
+        Arguments:
+
+            location: (float) Value of the CV where the wall will be located
+
+            kappa: (float) The force constant of the wall
```
Comment (tristan-j-wood) [inline mlptrain/sampling/plumed.py:232]: Add units

File: mlptrain/sampling/plumed.py [hunk 14]
```diff
@@ -215,6 +216,68 @@ def dof_sequence(self) -> str:
...
+                          ) -> None:
+        """
+        Attaches upper wall bias to the collective variable.
+
+        -----------------------------------------------------------------------
+        Arguments:
+
+            location: (float) Value of the CV where the wall will be located
+
+            kappa: (float) The force constant of the wall
```
Comment (tristan-j-wood) [inline mlptrain/sampling/plumed.py:263]: Add units

File: mlptrain/sampling/metadynamics.py [hunk 15]
```diff
@@ -715,43 +723,83 @@ def block_analysis(self,
         -----------------------------------------------------------------------
         Arguments:
 
+            energy_units: (str) Energy units to be used in plotting
```
Comment (tristan-j-wood) [inline mlptrain/sampling/metadynamics.py:726]: Could specify the allowed units

**Review by tristan-j-wood — APPROVED** [review]

**Pull request merged**